### PR TITLE
mark bad NISTs as bad in gregor auth

### PR DESCRIPTION
**Pull request by maxtaco**
- they were just being sent up repeatedly
- fail to find the original bug that mike found, but add more debugging

### CODE REVIEW

File: go/libkb/nist.go [hunk 1]
```diff
@@ -86,23 +86,41 @@ func (f *NISTFactory) NIST(ctx context.Context) (ret *NIST, err error) {
...
+		f.G().Log.CDebugf(ctx, "| NISTFactory#NIST: nil NIST, making new one")
+	} else if f.nist.DidFail() {
+		f.G().Log.CDebugf(ctx, "| NISTFactory#NIST: NIST previously failed, so we'll make a new one")
+	} else {
+		valid, until := f.nist.IsStillValid()
+		if valid {
+			f.G().Log.CDebugf(ctx, "| NISTFactory#NIST: returning existing NIST (expires conservatively in %s, expiresAt: %s)", until, f.nist.expiresAt)
+			makeNew = false
+		} else {
+			f.G().Log.CDebugf(ctx, "| NISTFactory#NIST: NIST expired (conservatively) %s ago, making a new one (expiresAt: %s)", -until, f.nist.expiresAt)
```
Comment (patrickxb) [inline go/libkb/nist.go]: why `-until`

**Comment by maxtaco** [comment]
Output looks like:

```
2018-04-26T13:27:17.586220-04:00 ▶ [DEBU keybase nist.go:98] 64c NISTFactory#NIST: returning existing NIST (expires conservatively in 3.350910359s, expiresAt: 2018-04-27 17:27:10.937428 -0400 EDT m=+100843.164397089) [tags:API=SA5qIjX4BEht]
2018-04-26T13:27:17.957900-04:00 ▶ [DEBU keybase nist.go:98] 67b NISTFactory#NIST: returning existing NIST (expires conservatively in 2.979213179s, expiresAt: 2018-04-27 17:27:10.937428 -0400 EDT m=+100843.164397089) [tags:API=6ZKUprvCzG8a]
2018-04-26T13:27:18.352608-04:00 ▶ [DEBU keybase nist.go:98] 6aa NISTFactory#NIST: returning existing NIST (expires conservatively in 2.584487552s, expiresAt: 2018-04-27 17:27:10.937428 -0400 EDT m=+100843.164397089) [tags:API=vt_pEJKe11UH]
2018-04-26T13:27:18.711850-04:00 ▶ [DEBU keybase nist.go:98] 6d9 NISTFactory#NIST: returning existing NIST (expires conservatively in 2.225229964s, expiresAt: 2018-04-27 17:27:10.937428 -0400 EDT m=+100843.164397089) [tags:API=GRBA_Ykn7-nd]
2018-04-26T13:27:22.998042-04:00 ▶ [DEBU keybase nist.go:101] 715 NISTFactory#NIST: NIST expired (conservatively) 2.061140708s ago, making a new one (expiresAt: 2018-04-27 17:27:10.937428 -0400 EDT m=+100843.164397089) [tags:API=OBiGIcaWvmdM]
2018-04-26T13:27:22.998088-04:00 ▶ [DEBU keybase nist.go:212] 716 + NIST#generate [tags:API=OBiGIcaWvmdM]
2018-04-26T13:27:22.998230-04:00 ▶ [DEBU keybase nist.go:107] 717 - NIST#generate -> ok [tags:API=OBiGIcaWvmdM]
2018-04-26T13:27:22.998277-04:00 ▶ [DEBU keybase nist.go:112] 718 NISTFactory#NIST: Installing new NIST; expiresAt: 2018-04-27 17:27:22.998127 -0400 EDT m=+100855.225641033 [tags:API=OBiGIcaWvmdM]
2018-04-26T13:27:23.402678-04:00 ▶ [DEBU keybase nist.go:98] 749 NISTFactory#NIST: returning existing NIST (expires conservatively in 9.595433177s, expiresAt: 2018-04-27 17:27:22.998127 -0400 EDT m=+100855.225641033) [tags:API=lrDgSvX9mej-]
2018-04-26T13:27:23.752421-04:00 ▶ [DEBU keybase nist.go:98] 778 NISTFactory#NIST: returning existing NIST (expires conservatively in 9.245674449s, expiresAt: 2018-04-27 17:27:22.998127 -0400 EDT m=+100855.225641033) [tags:API=WJ2Saw_j3aGb]
```

File: go/libkb/nist.go [hunk 2]
```diff
@@ -86,23 +86,41 @@ func (f *NISTFactory) NIST(ctx context.Context) (ret *NIST, err error) {
...
 }
 
-func (n *NIST) IsExpired() bool {
+func (n *NIST) IsStillValid() (bool, time.Duration) {
 	n.RLock()
 	defer n.RUnlock()
 	conservativeExpiration := n.expiresAt.Add(-nistExpirationMargin)
-	return !conservativeExpiration.After(n.G().Clock().Now())
+	diff := conservativeExpiration.Sub(n.G().Clock().Now())
+	return (diff > 0), diff
```
Comment (maxtaco) [inline go/libkb/nist.go]: note, this doesn't really change the logic or fix a bug. It just allows me to debug more. I can't figure out why the hell mike hit the bug he did.

File: go/service/gregor.go [hunk 3]
```diff
@@ -690,6 +690,7 @@ func (g *gregorHandler) OnConnect(ctx context.Context, conn *rpc.Connection,
 		if _, ok := err.(libkb.BadSessionError); ok {
 			g.chatLog.Debug(ctx, "bad session from SyncAll(): forcing session check on next attempt")
 			g.forceSessionCheck = true
+			nist.DidFail()
```
Comment (maxtaco) [inline go/service/gregor.go]: This did fix a bug, but only the bug that reconnect always failed after it failed once.  It doesn't fix the original failure.

File: go/libkb/nist.go (hunk 1 above)
Comment (maxtaco) [reply go/libkb/nist.go]: Creates output like:

```2018-04-26T13:27:22.998042-04:00 ▶ [DEBU keybase nist.go:101] 715 NISTFactory#NIST: NIST expired (conservatively) 2.061140708s ago, making a new one (expiresAt: 2018-04-27 17:27:10.937428 -0400 EDT m=+100843.164397089) [tags:API=OBiGIcaWvmdM]```

**Pull request merged**